rust/src/utils.rs

```diff
@@ -268,6 +268,94 @@ pub(crate) fn openat_to_dirfd(f: &openat::Dir) -> Result<cap_std::fs::Dir> {
     Ok(r)
 }
 
+/// Appends `part` to `base` in a way such that only characters that
+/// can be used in a D-Bus object path will be used. E.g. a character
+/// not in `[A-Z][a-z][0-9]_` will be escaped as `_HEX` where HEX is a
+/// two-digit hexadecimal number.
+///
+/// Note that his mapping is not bijective - e.g. you cannot go back
+/// to the original string.
+///
+/// # Arguments
+///
+/// * `base` - The base object path (without trailing '/').
+/// * `segment` - The UTF-8 string.
+pub(crate) fn append_to_object_path(base: &mut String, segment: impl AsRef<str>) {
+    for c in segment.as_ref().as_bytes() {
+        match *c {
+            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'_' => base.push(*c as char),
+            b'/' | b'-' => base.push('_'),
+            c => base.push_str(format!("_{c:02X}").as_str()),
+        }
+    }
+}
+
+/// Determines if a given string is a valid D-Bus object path.
+///
+/// A valid object path starts with / followed by zero or more sequences
+/// of characters separated by / characters. Each sequence must contain
+/// only the characters [A-Z][a-z][0-9]_. No sequence (including the one
+/// following the final / character) may be empty.
+pub(crate) fn is_object_path(path: &impl AsRef<str>) -> bool {
+    let path = path.as_ref().as_bytes();
+
+    if !matches!(path.first(), Some(b'/')) {
+        return false;
+    }
+
+    for pair in path.windows(2) {
+        match pair {
+            [b'/', b'/'] => return false,
+            [b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'_' | b'/', _] => {}
+            _ => return false,
+        }
+    }
+
+    // Window only yields full chunks
+    matches!(
+        path.last(),
+        Some(b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'_')
+    )
+}
+
+/// Appends `parts` to `base` in a way such that only characters that
+/// can be used in a D-Bus object path will be used. E.g. a character
+/// not in `[A-Z][a-z][0-9]_` will be escaped as `_HEX` where HEX is a
+/// two-digit hexadecimal number.
+///
+/// Note that his mapping is not bijective - e.g. you cannot go back
+/// to the original string.
+///
+/// # Arguments
+///
+/// * `base` - The base object path (without trailing '/').
+/// * `segments` - The UTF-8 strings.
+fn generate_object_path_impl<S, P>(base: impl std::string::ToString, segments: P) -> Result<String>
+where
+    S: AsRef<str>,
+    P: std::iter::IntoIterator<Item = S>,
+{
+    let mut base = base.to_string();
+    if !is_object_path(&base) {
+        return Err(anyhow::anyhow!("Value is not an object path: {base}"));
+    }
+
+    for segment in segments {
+        let segment = segment.as_ref();
+        if segment.len() == 0 {
+            return Err(anyhow::anyhow!("Cannot append empty part: {base}"));
+        }
+        base.push('/');
+        append_to_object_path(&mut base, segment);
+    }
+
+    return Ok(base);
+}
+
+pub(crate) fn generate_object_path(base: &str, segments: &[&str]) -> CxxResult<String> {
```

Comment:
It looks like all consumers are actually using a single-entry `segments`.
So perhaps we can simplify this and directly take `(base: &str, osname: &str)` as input arguments.